hack/generator/pkg/astmodel/type_walker.go

```diff
@@ -0,0 +1,195 @@
+
+// TODO: it's awkward to have so much configuration on this thing (3 separate funcs that apply at different places in the walking process?)
+// TODO: but unsure if there's a better way... bring it up in code review.
+
+// TypeWalker performs a depth first search across the types provided, applying the visitor to each TypeDefinition.
+// MakeContextFunc is called before each visit, and AfterVisitFunc is called after each visit. WalkCycle is called
+// if a cycle is detected.
+type TypeWalker struct {
+	allTypes Types
+	visitor  TypeVisitor
+
+	// MakeContextFunc is called before a type is visited.
+	MakeContextFunc func(it TypeName, ctx interface{}) (interface{}, error)
+	// AfterVisitFunc is called after the type walker has applied the visitor to a TypeDefinition.
+	AfterVisitFunc func(original TypeDefinition, updated TypeDefinition, ctx interface{}) (TypeDefinition, error)
+	// WalkCycle is called if a cycle is detected. It allows configurable behavior for how to handle cycles.
+	WalkCycle func(def TypeDefinition, ctx interface{}) (TypeName, error)
+
+	state                   typeWalkerState
+	originalVisitTypeName   func(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error)
+	originalVisitObjectType func(this *TypeVisitor, it *ObjectType, ctx interface{}) (Type, error)
+}
+
+type typeWalkerState struct {
+	result     Types
+	processing map[TypeName]struct{}
+}
+
+// NewTypeWalker returns a TypeWalker.
+// The provided visitor VisitTypeName function must return a TypeName and VisitObjectType must return an ObjectType or calls to
+// Walk will panic.
+func NewTypeWalker(allTypes Types, visitor TypeVisitor) *TypeWalker {
+	typeWalker := TypeWalker{
+		allTypes: allTypes,
+	}
+	typeWalker.originalVisitTypeName = visitor.VisitTypeName
+	typeWalker.originalVisitObjectType = visitor.VisitObjectType
+
+	// visitor is a copy - modifications won't impact passed visitor
+	visitor.VisitTypeName = typeWalker.visitTypeName
+	visitor.VisitObjectType = typeWalker.visitObjectType
+
+	typeWalker.visitor = visitor
+	typeWalker.AfterVisitFunc = DefaultAfterVisit
+	typeWalker.MakeContextFunc = IdentityMakeContext
+	typeWalker.WalkCycle = IdentityWalkCycle
+
+	return &typeWalker
+}
+
+func (t *TypeWalker) visitTypeName(this *TypeVisitor, it TypeName, ctx interface{}) (Type, error) {
+	updatedCtx, err := t.MakeContextFunc(it, ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	visitedTypeName, err := t.originalVisitTypeName(this, it, updatedCtx)
+	if err != nil {
+		return nil, err
+	}
+	var ok bool
+	it, ok = visitedTypeName.(TypeName)
+	if !ok {
+		panic(fmt.Sprintf("TypeWalker visitor VisitTypeName must return a TypeName, instead returned %T", visitedTypeName))
+	}
+
+	def, ok := t.allTypes[it]
+	if !ok {
+		return nil, errors.Errorf("couldn't find type %q", it)
+	}
+
+	// Prevent loops by bypassing this type if it's currently being processed. The processing
+	// slice is basically the "path" taken to get to the current type.
```

Comment:
Looks like this comment was written before processing was converted to a map?

Reply:
Fixed